src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -370,36 +370,59 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
         {
             ImmutableArray<byte> data = default;
             int elementCount = -1;
-            TypeSymbol elementType = null;
+            MethodSymbol createSpan = null;
+            avoidInPlace = false;
 
             if (!_module.SupportsPrivateImplClass)
             {
                 return false;
             }
 
-            var ctor = ((MethodSymbol)this._module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer));
+            var ctor = (MethodSymbol)_module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer);
             if (ctor == null)
             {
                 return false;
             }
 
             if (wrappedExpression is BoundArrayCreation ac)
             {
-                var arrayType = (ArrayTypeSymbol)ac.Type;
-                elementType = arrayType.ElementType.EnumUnderlyingTypeOrSelf();
-
-                // NB: we cannot use this approach for element types larger than one byte
-                //     the issue is that metadata stores blobs in little-endian format
-                //     so anything that is larger than one byte will be incorrect on a big-endian machine
-                //     With additional runtime support it might be possible, but not yet.
-                //     See: https://github.com/dotnet/corefx/issues/26948 for more details
-                if (elementType.SpecialType.SizeInBytes() != 1)
+                var elementType = ((ArrayTypeSymbol)ac.Type).ElementType;
+
+                switch (elementType.EnumUnderlyingTypeOrSelf().SpecialType)
                 {
-                    return false;
+                    case SpecialType.System_SByte:
+                    case SpecialType.System_Byte:
+                    case SpecialType.System_Boolean:
+                        // For primitives that are a single byte in size, a span can point directly to the blob.
+                        break;
+
+                    case SpecialType.System_Int16:
+                    case SpecialType.System_UInt16:
+                    case SpecialType.System_Int32:
+                    case SpecialType.System_UInt32:
+                    case SpecialType.System_Int64:
+                    case SpecialType.System_UInt64:
+                    case SpecialType.System_Char:
+                    case SpecialType.System_Single:
+                    case SpecialType.System_Double:
+                        // For primitives that are larger than a single byte, endianness needs to be considered,
+                        // which requires then using the runtime's RuntimeHelpers.CreateSpan if it's available.
+                        createSpan = (MethodSymbol)_module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle);
+                        if (createSpan is null)
```

Comment:
>createSpan is null [](http://example.com/codeflow?start=28&length=18)

In general, not null value doesn't mean the member is good to use. We should check if there are still errors associated with it. For example, check ```GetWellKnownTypeMember``` helpers in the Binder.

Reply:
There's existing code a few lines before this that's doing:
```C#
var ctor = ((MethodSymbol)this._module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer));
```
Is that also wrong?

Reply:
> There's existing code a few lines before this that's doing:
> ...
> without checking diagnostics. Is that also wrong? 

That code also performs an incomplete check for a usage of a well known member.

>  I was just copying what's already being done in this file.

 I understand that, perhaps, the likelihood to find a method that is still "bad" is extremely low for these helpers, but it is better to always perform full validation. One of the reasons is the fact that code is often copied for other helpers and between compiler layers.